### ISSUE TITLE
ossec-agent crashing when starting syscheck on Windows 2019 and Windows 2012 R2

### ISSUE THREAD

**Issue by hanschl** [wazuh/wazuh#6456]
|Wazuh version|Component|Install type|Install method|Platform|
|3.12.0 - 4.0.0|syscheckd|Agent|MSI|Windows 2019 & Windows 2012 R2|

On 2 of our Windows Servers (We have around 40 connected to wazuh) ossec-agent is crashing as soon as syscheck ist starting. It happens once the frequency in the configuration is reached (in our case 8 hours) or if we trigger the syscheck from wazuh app but not after the restart of the service.

We get the following Event when the agent is crashing:

On Windows Server 2012 R2:

```
10/29/2020 09:50:44 AM
LogName=Application
SourceName=Application Error
EventCode=1000
EventType=2
Type=Error
ComputerName=***
TaskCategory=Application Crashing Events
OpCode=Info
RecordNumber=202816
Keywords=Classic
Message=Faulting application name: ossec-agent.exe, version: 0.0.0.0, time stamp: 0x5f918392
Faulting module name: msvcrt.dll, version: 7.0.9600.17415, time stamp: 0x54504b2e
Exception code: 0xc0000005
Fault offset: 0x0001b3d4
Faulting process id: 0xeac
Faulting application start time: 0x01d6add049c66778
Faulting application path: C:\Program Files (x86)\ossec-agent\ossec-agent.exe
Faulting module path: C:\Windows\SYSTEM32\msvcrt.dll
Report Id: d448736b-19c3-11eb-8117-74e6e2fe03f9
Faulting package full name: 
Faulting package-relative application ID:
```

On Windows Server 2019:

```
10/29/2020 10:14:49 AM
LogName=Application
SourceName=Application Error
EventCode=1000
EventType=2
Type=Error
ComputerName=***
TaskCategory=Application Crashing Events
OpCode=Info
RecordNumber=261246
Keywords=Classic
Message=Faulting application name: ossec-agent.exe, version: 0.0.0.0, time stamp: 0x5f918392
Faulting module name: msvcrt.dll, version: 7.0.17763.475, time stamp: 0xba51b082
Exception code: 0xc0000005
Fault offset: 0x0008ab00
Faulting process id: 0x1f98
Faulting application start time: 0x01d6adcb90f72380
Faulting application path: C:\Program Files (x86)\ossec-agent\ossec-agent.exe
Faulting module path: C:\Windows\System32\msvcrt.dll
Report Id: 3c383687-c556-4394-b0dc-f18ccfee1891
Faulting package full name: 
Faulting package-relative application ID:
```
I've already found out that the last version of wazuh that isn't affected by this is 3.11.4 any newer version installed on these 2 specific servers is causing the crash. 

Additionally on the windows 2019 Server we get events like this every 2-5 minutes:
```
10/29/2020 10:20:06 AM
LogName=Application
SourceName=Windows Error Reporting
EventCode=1001
EventType=4
Type=Information
ComputerName=***
TaskCategory=The operation completed successfully.
OpCode=Info
RecordNumber=261248
Keywords=Classic
Message=Fault bucket , type 0
Event Name: APPCRASH
Response: Not available
Cab Id: 0

Problem signature:
P1: ossec-agent.exe
P2: 0.0.0.0
P3: 5f918392
P4: msvcrt.dll
P5: 7.0.17763.475
P6: ba51b082
P7: c0000005
P8: 0008ab00
P9: 
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER2CAA.tmp.dmp
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER3100.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER3140.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER313E.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER315E.tmp.txt
\\?\C:\Windows\Temp\WER94.tmp.appcompat.txt
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportQueue\AppCrash_ossec-agent.exe_fc5fb0f376b145747a774a47c1e943a01742d_d06cee23_cab_3aa8010d\memory.hdmp

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportQueue\AppCrash_ossec-agent.exe_fc5fb0f376b145747a774a47c1e943a01742d_d06cee23_cab_3aa8010d

Analysis symbol: 
Rechecking for solution: 0
Report Id: 3c383687-c556-4394-b0dc-f18ccfee1891
Report Status: 100
Hashed bucket: 
Cab Guid: 0
```

Could this be a bug in the ossec-agent?

**Comment by Molter73** [comment]
Hi @hanschl 

We recently fixed [this issue](https://github.com/wazuh/wazuh/issues/6055), where a file with no permissions on Windows would cause a crash similar to the one you describe, an easy way to check if this is your problem would be to add `check_perm="no"` to your `<directories>` configuration and see if it fixes the crash. If it does, you can narrow down where the file with no permissions is located by removing that attribute from your configurations one at a time.

If the above did not fix the crash, I would ask that you share with us the configuration on those agents as reported by [this API call](https://documentation.wazuh.com/4.0/user-manual/api/reference.html#operation/api.controllers.agents_controller.get_agent_config), in this case the fields `component` and  `configuration` would be `syscheck` for both.

Additionally, if you could put the agents in debug mode, cause the crash and share the log with us, it would be of great use.

**Comment by hanschl** [comment]
Hi @Molter73 

thanks for the tip. Using check_perm="no" did help with the crashing agent. We are currently trying to find out which file exactly is causing the problem. In which version of the agent will the problem be fixed?

I will attach the debug log as soon as we did the testing.

**Comment by hanschl** [comment]
[ossec.log](https://github.com/wazuh/wazuh/files/5473773/ossec.log)
I have attached an extract of the ossec.log and it is crashing when it's trying to read the file **c:\programdata\microsoft\crypto\rsa\machinekeys\148db85c53b64b028983057aa4ff7aa6_4f828323-1af7-402b-82b0-fbb830762272** which doesn't have any file permissions
![image](https://user-images.githubusercontent.com/68237718/97849560-04e9e480-1cf3-11eb-8c3d-95c4e33e8122.png)

**Comment by Molter73** [comment]
Hi @hanschl 

Glad to hear you could find the source of this problem, the fix will be available starting with version 4.1.0.

I will keep the issue open for a couple of days in case you need help with anything related with it.
